openmdao/core/group.py

```diff
@@ -711,70 +722,52 @@ def _sub_apply_linear_wrapper(self, system, mode, vois, ls_inputs=None,
         gs_outputs : dict, optional
             Linear Gauss-Siedel can limit the outputs when calling apply.
         """
+        states = system.states
+        force_fd = system.fd_options['force_fd']
 
         for voi in vois:
+            if not self._relevance.is_relevant_system(voi, system):
+                continue
 
             dresids = system.drmat[voi]
             dunknowns = system.dumat[voi]
             dparams = system.dpmat[voi]
+            gsouts = None if gs_outputs is None else gs_outputs[voi]
 
             # Linear GS imposes a stricter requirement on whether or not to run.
-            abs_inputs = {meta['pathname'] for meta in itervalues(dparams)}
+            abs_inputs = system._abs_inputs[voi]
+            do_apply = ls_inputs[voi] is None or (abs_inputs and abs_inputs.intersection(ls_inputs[voi]))
 
             # Forward Mode
             if mode == 'fwd':
 
                 dresids.vec[:] = 0.0
 
-                if ls_inputs[voi] is None or abs_inputs.intersection(ls_inputs[voi]):
-
-                    # Speedhack, don't call component's derivatives if
-                    # incoming vector is zero.
-                    nonzero = False
-                    for key in system._get_fd_params():
-                        try:
-                            value = dparams.flat[key]
```

Comment:
the pymission optimization was faster without the zero checks, so I took them out.  Maybe we need to benchmark against some other models, like CADRE for example, to see if taking them out has an impact.